### ISSUE TITLE
new ServiceClient "hanging"

### ISSUE THREAD

**Issue by mjfpalmer** [microsoft/PowerPlatform-DataverseServiceClient#213]
### Discussed in https://github.com/microsoft/PowerPlatform-DataverseServiceClient/discussions/212

I'm creating a .Net 6.0 Blazor Server project which references version  0.5.10 of the DataverseServiceClient and injects a Singleton wrapper class containing an instance of a ServiceClient for use in Controllers and Razor pages/components through DI.

If I instantiate an instance of ServiceClient in Program.cs _before_ the application starts running, it connects fine, IsReady is true and the ServiceClient can retrieve/create/etc. This confirms that my CRM connection string is correct. The app connects to CRM using the ClientSecret AuthType.

If I attempt to use the class after the app has started running, when it attempts to instantiate a new ServiceClient using the constructor below, then the application hangs on this line forever without ever throwing an exception or running past the instantiation.
`this.ServiceClient = new ServiceClient(organisationURL, clientId, clientSecret, true);`

Applying the suggested fix from #113 does allow me to get around the issue, but without it, I'm still getting the "hang".

Happy to provide more info if needed...

**Comment by mjfpalmer** [comment]
If it helps, I'll gladly provide a sample project to highlight the error - just let me know if that's needed.

**Comment by MattB-msft** [comment]
That would be helpfull, feel free to link a git hub repro and we will have our folks take a look.

**Comment by mjfpalmer** [comment]
Sure. Here's a _very_ barebones example project https://github.com/mjfpalmer/blazor-server-service-client-example-1 that shows the issue.

https://github.com/mjfpalmer/blazor-server-service-client-example-1/commit/ac085aa0b75d62a1b79d04b452e7a3c25802eb8e contains the standard .Net 6.0 Blazor Server Template.
https://github.com/mjfpalmer/blazor-server-service-client-example-1/commit/fc549e5162e6d677be4f7ccf23d45bc59388fd7d contains the addition of the ServiceClient.

To run the app, provide a working Dataverse connection string in BlazorServerServiceClientExample/Contexts/ServiceClientContext.cs.
You'll see then that the code never reaches line 11 of ServiceClientContext.cs when a successful connection happens on line 10. The code does proceed past line 10 if the connection fails FYI.

**Comment by petrochuk** [comment]
Thank you for compact repro. Dataverse ServiceClient constructor performs some async actions and it is getting deadlocked when used with dependency injection "inject" which runs very early in page construction. If you move context creation (which creates ServiceClient) to OnInitialized method your code will work fine. 


Here is sample page from your example:
(Notice I call ServiceClientContext.ServiceClient.IsReady to verify)

```HTML+Razor
@page "/"
<!-- inject BlazorServerServiceClientExample.Contexts.ServiceClientContext ServiceClientContext -->
@inject IServiceProvider ServiceProvider 

<PageTitle>Index</PageTitle>

<h1>Hello, world!</h1>

Welcome to your new app.

<SurveyPrompt Title="How is Blazor working for you?" />

@if (ServiceClientContext.ServiceClient.IsReady)
{
    <text>Dataverse is ready</text>
}

@code {
    BlazorServerServiceClientExample.Contexts.ServiceClientContext ServiceClientContext;

    protected override void OnInitialized()
    {
        ServiceClientContext = ServiceProvider.GetRequiredService<BlazorServerServiceClientExample.Contexts.ServiceClientContext>();
    }
}
```

**Comment by mjfpalmer** [comment]
Thanks @petrochuk. 

My issue with the OnInitialized approach is that it requires me to place this code on every page and results in a generally poorly architected solution that is not able to make use of good dependency injection, so I'd really rather that the actual problem be addressed to allow us to inject the ServiceClient. @MattB-msft did mention that this was planned in https://github.com/microsoft/PowerPlatform-DataverseServiceClient/discussions/187 so I guess the issue would have to be fixed before then?

FYI, I already have several workarounds, including the one below that actually allows me to use the ServiceClient in injected services without having to initialize it on every page.

```
  /// <summary>
  /// The DynamicsContext class
  /// </summary>
  public class DynamicsContext : IDynamicsContext, IDisposable
  {
    private ServiceClient? serviceClient;

    /// <summary>
    /// Initializes a new instance of the <see cref="DynamicsContext"/> class.
    /// </summary>
    /// <param name="dynamicsSettings">The Dynamics settings.</param>
    public DynamicsContext(IOptions<DynamicsSettings> dynamicsSettings)
    {
      this.DynamicsSettings = dynamicsSettings.Value;
    }

    private DynamicsSettings DynamicsSettings { get; set; }

    /// <summary>
    /// Gets the service client.
    /// </summary>
    private ServiceClient ServiceClient
    {
      get
      {
        if (this.serviceClient == null)
        {
          Task.Run(async () => { this.serviceClient = new ServiceClient(this.DynamicsSettings.OrganisationURL, this.DynamicsSettings.ClientId, ServiceClient.MakeSecureString(this.DynamicsSettings.ClientSecret), true); }).Wait();

          if (!this.serviceClient.IsReady)
          {
            throw this.serviceClient.LastException;
          }
        }

        return this.serviceClient;
      }
    }

...
```

**Comment by petrochuk** [comment]
One more solution is to use Lazy initialization which will solve need for OnInitialized

```HTML+Razor
namespace BlazorServerServiceClientExample.Contexts
{
    using Microsoft.PowerPlatform.Dataverse.Client;

    public class ServiceClientContext
    {
        public ServiceClientContext()
        {
        }

        public Lazy<ServiceClient> ServiceClient { get; } = new Lazy<ServiceClient>(() => new ServiceClient("<connection string here>");
    }
}
```

**Comment by mjfpalmer** [comment]
Thanks @petrochuk! Yeah, the DynamicsContext solution above does use a lazy initialization pattern in the ServiceClient property, without actually using the Lazy object. I'll look into using Lazy and see if that works and allows me to avoid the Task.Run(...).Wait() approach which I don't like at all.

**Comment by mjfpalmer** [comment]
@petrochuk I've tried out the lazy initialization approach, and unfortunately the deadlock still seems to occur.

See https://github.com/mjfpalmer/blazor-server-service-client-example-1/tree/test-lazy-initialization for a branch of the example project with the Lazy initialization implemented with the deadlock still happening.

I'm ok to stick with my currently working workaround (as implemented on https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/213#issuecomment-990650214) for now, as long as I know that the actual deadlock issue is being investigated and will eventually be solved.

**Comment by TclasenITVT** [comment]
We also experienced a hanging new ServiceClient in the context of a WPF App (Net Framework 4.7.2).

**Comment by sharpt00th** [comment]
I was updating a .Net Framework 4.7.2 Web API project recently to use the Dataverse ServiceClient instead of CrmServiceClient and encountered problems that align with the issue described here, albeit not one of initialization timing but I'm seeing the exact same symptom of a hanging ServiceClient instantiation. I remembered creating a sample Web API project to test the Dataverse ServiceClient functionality back when the package was at version 0.4.21 - this project targets .Net Framework 4.8 and still works as expected so I did some further investigation to see if I could reliably reproduce the issue and it seems I can.

I created two initially identical Web API projects in Visual Studio 2022 using the default Web API template and targeting .Net Framework 4.8.
The first project contains a simple controller and action that initialises a ServiceClient connection using ClientID\Secret authentication. It references Microsoft.PowerPlatform.Dataverse.Client version **0.4.21**. No other packages have been updated. When called with valid working credentials the action returns as expected.
The second project contains the same controller and action. It references Microsoft.PowerPlatform.Dataverse.Client version **0.6.1** and all other packages needing updating have been updated. When called with valid working credentials the action hangs on the line containing "new ServiceClient(..."
The action is written as follows:
```
[HttpPost]
public IHttpActionResult Call(TestConnectionRequest request)
{
   try
   {
      var client = new ServiceClient(new Uri(request.InstanceUrl), request.ClientId, request.ClientSecret, true);
      return Content(System.Net.HttpStatusCode.OK, client);
   }
   catch (Exception ex)
   {
      return BadRequest(ex.Message);
   }
}
```

Both projects are in my github repo [here](https://github.com/sharpt00th/sharpiversal) where SimpleDotNetFourEightWebApi is the "first project" and UpdatedDotNetFourEightWebApi is the "second project". I concede it's entirely possible this problem doesn't relate to OP's issue but the symptom is identical and I noted the most recent comment from TclasenITVT describes the same symptom with a different project type.
